File: typescript/packages/langchain/src/tools.ts
import { tool } from '@langchain/core/tools';
import { z } from 'zod';
import {
  DISCOVER_DESCRIPTION,
  DISCOVER_QUERY_PARAM,
  INTERPRET_DESCRIPTION,
  INTERPRET_ITEMS_PARAM,
  INTERPRET_QUERY_PARAM,
  SEARCH_DESCRIPTION,
  SEARCH_QUERY_PARAM,
} from '@copass/config';
import type { CopassClient, SearchPreset, WindowLike } from '@copass/core';

export interface CopassToolsOptions {
  /** An authenticated `@copass/core` client. */
  client: CopassClient;
  /** Sandbox all retrieval runs against. */
  sandbox_id: string;
  /** Optional project scoping for retrieval calls. */
  project_id?: string;
  /**
   * Optional conversation window — a {@link WindowLike} (typically a
   * `ContextWindow` from `client.contextWindow.create()`). When provided,
   * every retrieval call is automatically window-aware.
   */
  window?: WindowLike;
  /**
   * Preset for `discover`, `interpret`, and `search`. Defaults to
   * `"copass/copass_1.0"`. Under `"copass/copass_2.0"` discover items
   * carry `subgraph` (pre-rendered ASCII tree) and `matched_query_nodes`
   * fields. Append `":thinking"` (e.g. `"copass/copass_2.0:thinking"`)
   * to enable task decomposition before retrieval on `search`.
   */
  preset?: SearchPreset;
}

/**
 * Return Copass retrieval as a set of LangChain tool objects (built with
 * `tool()` from `@langchain/core/tools`).
 *
 * Agent-framework-neutral shape: the LLM chooses whether to surface a menu
 * (`discover`), drill into specific items (`interpret`), or get a
 * synthesized answer in one shot (`search`).
 *
 * @example
 * ```ts
 * import { ChatAnthropic } from '@langchain/anthropic';
 * import { createReactAgent } from '@langchain/langgraph/prebuilt';
 * import { copassTools } from '@copass/langchain';
 *
 * const window = await copass.contextWindow.create({ sandbox_id });
 * const tools = copassTools({ client: copass, sandbox_id, window });
 *
 * const agent = createReactAgent({
 *   llm: new ChatAnthropic({ model: 'claude-opus-4-7' }),
 *   tools: [tools.discover, tools.interpret, tools.search],
 * });
 *
 * const result = await agent.invoke({
 *   messages: [{ role: 'user', content: 'why is checkout flaky?' }],
 * });
 * ```
 */
export function copassTools(options: CopassToolsOptions) {
  const { client, sandbox_id, project_id, window, preset = 'copass/copass_1.0' } = options;

  const discover = tool(
    async ({ query }: { query: string }) => {
      const response = await client.retrieval.discover(sandbox_id, {
        query,
        project_id,
        window,
        preset,
      });
      return {
        header: response.header,
        // Project the v2 fields (`subgraph` + `matched_query_nodes`)
        // alongside the v1 fields. Populated only under
        // `copass/copass_2.0` (or its `copass/2.0` alias); `null` under
        // v1 — agents can ignore them when not present.
        items: response.items.map((item) => ({
          score: item.score,
          summary: item.summary,
          canonical_ids: item.canonical_ids,
          subgraph: item.subgraph ?? null,
          matched_query_nodes: item.matched_query_nodes ?? null,
        })),
        next_steps: response.next_steps,
      };
    },
    {
      name: 'discover',
      description: DISCOVER_DESCRIPTION,
      schema: z.object({
        query: z.string().describe(DISCOVER_QUERY_PARAM),
      }),
    },
  );

  const interpret = tool(
    async ({ query, items }: { query: string; items: string[][] }) => {
      const response = await client.retrieval.interpret(sandbox_id, {
        query,
        items,
        project_id,
        window,
        preset,
      });
      return { brief: response.brief };
    },
    {
      name: 'interpret',
      description: INTERPRET_DESCRIPTION,
      schema: z.object({
        query: z.string().describe(INTERPRET_QUERY_PARAM),
        items: z.array(z.array(z.string())).min(1).describe(INTERPRET_ITEMS_PARAM),
      }),
    },
  );

  const search = tool(
    async ({ query }: { query: string }) => {
      const response = await client.retrieval.search(sandbox_id, {
        query,
        project_id,
        window,
        preset,
      });
      return { answer: response.answer };
    },
    {
      name: 'search',
      description: SEARCH_DESCRIPTION,
      schema: z.object({
        query: z.string().describe(SEARCH_QUERY_PARAM),
      }),
    },
  );

  return { discover, interpret, search };
}

File: python/copass-pydantic-ai/src/copass_pydantic_ai/tools.py
"""Pydantic AI tool adapters for Copass retrieval.

``copass_tools(...)`` returns a tuple of three async callables — drop them
straight into ``Agent(tools=[...])``. Each function has a type-hinted
signature; its ``__doc__`` is set from the canonical descriptions in
``_strings`` (which mirrors the ``@copass/config`` TypeScript package), so
the schema + description Pydantic AI generates matches what every other
adapter in the copass surfaces.
"""

from __future__ import annotations

from typing import Any, Callable, Optional

from ._strings import (
    DISCOVER_DESCRIPTION,
    INTERPRET_DESCRIPTION,
    SEARCH_DESCRIPTION,
)
from .client import CopassRetrievalClient
from .types import SearchPreset, WindowLike


def copass_tools(
    *,
    client: CopassRetrievalClient,
    sandbox_id: str,
    project_id: Optional[str] = None,
    window: Optional[WindowLike] = None,
    preset: SearchPreset = "copass/copass_1.0",
) -> tuple[Callable[..., Any], Callable[..., Any], Callable[..., Any]]:
    """Return Copass retrieval as three Pydantic AI-compatible tool callables.

    Args:
        client: A :class:`CopassRetrievalClient` for the target sandbox.
        sandbox_id: Sandbox all retrieval runs against.
        project_id: Optional project scoping.
        window: Optional :class:`WindowLike` (e.g. a Context Window handle).
            When provided, every retrieval call is window-aware.
        preset: Preset for ``discover`` / ``interpret`` / ``search``.
            Defaults to ``"copass/copass_1.0"``. Under
            ``"copass/copass_2.0"`` discover items carry ``subgraph``
            (pre-rendered ASCII tree) and ``matched_query_nodes``
            fields. Append ``":thinking"`` (e.g.
            ``"copass/copass_2.0:thinking"``) to enable task
            decomposition before retrieval on ``search``.

    Returns:
        ``(discover, interpret, search)`` — three async callables, ready to
        pass directly to ``pydantic_ai.Agent(tools=[...])``.

    Example::

        from pydantic_ai import Agent
        from copass_pydantic_ai import CopassRetrievalClient, copass_tools

        client = CopassRetrievalClient(api_url="...", api_key="olk_...")
        discover, interpret, search = copass_tools(
            client=client,
            sandbox_id="sandbox_abc",
        )

        agent = Agent("anthropic:claude-opus-4-7", tools=[discover, interpret, search])
        result = await agent.run("why is checkout flaky?")
    """

    async def discover(query: str) -> dict[str, Any]:
        response = await client.discover(
            sandbox_id,
            query=query,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {
            "header": response.get("header"),
            "items": [
                # Project the v2 fields (``subgraph`` + ``matched_query_nodes``)
                # alongside the v1 fields. Populated only under
                # ``copass/copass_2.0`` (or its ``copass/2.0`` alias);
                # ``None`` under v1.
                {
                    "score": item.get("score"),
                    "summary": item.get("summary"),
                    "canonical_ids": item.get("canonical_ids", []),
                    "subgraph": item.get("subgraph"),
                    "matched_query_nodes": item.get("matched_query_nodes"),
                }
                for item in response.get("items", [])
            ],
            "next_steps": response.get("next_steps"),
        }

    async def interpret(query: str, items: list[list[str]]) -> dict[str, Any]:
        response = await client.interpret(
            sandbox_id,
            query=query,
            items=items,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {"brief": response.get("brief")}

    async def search(query: str) -> dict[str, Any]:
        response = await client.search(
            sandbox_id,
            query=query,
            project_id=project_id,
            window=window,
            preset=preset,
        )
        return {"answer": response.get("answer")}

    # Pydantic AI reads each tool's description from `fn.__doc__` at agent
    # construction. Setting it here (rather than inlining as source-level
    # docstrings) pins the LLM-facing copy to the shared `_strings` module
    # so the Python package can never drift from @copass/config.
    discover.__doc__ = DISCOVER_DESCRIPTION
    interpret.__doc__ = INTERPRET_DESCRIPTION
    search.__doc__ = SEARCH_DESCRIPTION

    return discover, interpret, search
